envoy-control-tests/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/permissions/ClientNameTrustedHeaderTest.kt

```diff
@@ -0,0 +1,180 @@
+        """.trimIndent()
+
+        @JvmField
+        @RegisterExtension
+        val envoy = EnvoyExtension(envoyControl, service, Echo1EnvoyAuthConfig.copy(configOverride = proxySettings))
+
+        // language=yaml
+        private val echoClientsConfig = """
+            node:
+              metadata:
+                proxy_settings:
+                  outgoing:
+                    dependencies:
+                      - service: "echo"
+        """.trimIndent()
+
+        @JvmField
+        @RegisterExtension
+        val envoy2 = EnvoyExtension(envoyControl, service, Echo2EnvoyAuthConfig.copy(configOverride = echoClientsConfig))
+
+        val echo4EnvoyAuthConfig = Echo1EnvoyAuthConfig.copy(
+            serviceName = "echo4",
+            certificateChain = "/app/fullchain_echo4.pem",
+            privateKey = "/app/privkey_echo4.pem",
+            configOverride = echoClientsConfig
+        )
+
+        @JvmField
+        @RegisterExtension
+        val envoy4MultipleSANs = EnvoyExtension(envoyControl, service, echo4EnvoyAuthConfig)
+
+        val echo5EnvoyAuthConfig = Echo1EnvoyAuthConfig.copy(
+            serviceName = "echo5",
+            certificateChain = "/app/fullchain_echo5.pem",
+            privateKey = "/app/privkey_echo5.pem",
+            configOverride = echoClientsConfig
+        )
+
+        @JvmField
+        @RegisterExtension
+        val envoy5InvalidSANs = EnvoyExtension(envoyControl, service, echo5EnvoyAuthConfig)
+    }
+
+    @BeforeEach
+    fun beforeEach() {
+        consul.server.operations.registerService(
+            id = "echo",
+            name = "echo",
+            address = envoy.container.ipAddress(),
+            port = EnvoyContainer.INGRESS_LISTENER_CONTAINER_PORT,
+            tags = listOf("mtls:enabled")
+        )
+        waitForEnvoysInitialized()
+    }
+
+    private fun waitForEnvoysInitialized() {
+        untilAsserted(wait = Duration.ofSeconds(20)) {
+            assertThat(envoy2.container.admin().isEndpointHealthy("echo", envoy.container.ipAddress())).isTrue()
+            assertThat(envoy4MultipleSANs.container.admin().isEndpointHealthy("echo", envoy.container.ipAddress())).isTrue()
+            assertThat(envoy5InvalidSANs.container.admin().isEndpointHealthy("echo", envoy.container.ipAddress())).isTrue()
+        }
+    }
+
+    @Test
+    fun `should always remove "x-client-name-trusted" header on every envoy ingress request`() {
+        // when
+        val response = envoy2.ingressOperations.callLocalService(
+            "/endpoint",
+            Headers.of(mapOf("x-client-name-trusted" to "fake-service"))
+        )
+
+        // then
+        assertThat(response).isOk()
+        HttpsEchoResponse(response).also {
+            assertThat(it).isFrom(service.container())
+            assertThat(it.requestHeaders["x-client-name-trusted"]).isNull()
+        }
+    }
+
+    @Test
+    fun `should add "x-client-name-trusted" header on envoy ingress request`() {
```

Comment:
This and the test above suggest two exclusive events ("should remove" and "should add"). Maybe we should specify that we remove the header from `ingress` and add it to `local_service`.

Reply:
I can change the names respectively to:
* `should always remove "x-client-name-trusted" header from client ingress request`
* `should add "x-client-name-trusted" header to ingress request to local service`
LMK if I understood you correctly.

Reply:
changed, hope it was what you meant.